src/session_manager.rs

```diff
@@ -0,0 +1,1055 @@
+                                functions::CheckDatabaseEncryptionKey::new()
+                                    .encryption_key(encryption_key)
+                                    .send(client_id),
+                                |result| async {
+                                    if let Err(e) = result {
+                                        panic!("Error on sending encryption key: {:?}", e);
+                                    }
+                                },
+                            );
+                        }
+                        AuthorizationState::Ready => {
+                            self_.sessions_handled.set(self_.sessions_handled.get() + 1);
+
+                            self.add_logged_in_session(client_id, database_info, is_active);
+
+                            if self_.initial_sessions_to_handle.get()
+                                == self_.sessions_handled.get()
+                            {
+                                // Just be sure to switch to the sessions after all initial sessions
+                                // have been handled.
+                                self.switch_to_sessions(None);
+                            }
+                        }
+                        other => {
+                            self_.sessions_handled.set(self_.sessions_handled.get() + 1);
+
+                            if self_.initial_sessions_to_handle.get() == 1 {
+                                if APPLICATION_OPTS.get().unwrap().test_dc
+                                    != database_info.use_test_dc
+                                {
+                                    self.tear_down_client(client_id, || {}, || {});
+                                    self.add_new_session(!database_info.use_test_dc);
+                                } else {
+                                    // Overwrite ClientInfo.
+                                    self_.clients.borrow_mut().insert(
+                                        client_id,
+                                        ClientInfo::Auth {
+                                            maybe_authorized: false,
+                                            is_active,
+                                            database_info: database_info.clone(),
+                                        },
+                                    );
+
+                                    self_.login.login_client(client_id, database_info);
+
+                                    self_
+                                        .login
+                                        .set_authorization_state(update.authorization_state);
+                                }
+                            } else {
+                                match other {
+                                    AuthorizationState::LoggingOut => {
+                                        log::warn!(
+                                            "Unexpected authorization state '{:?}' for previously \
```

Comment:
Is it really unexpected? If there's even an authorization state just for that I think it's indeed expected.

Reply:
Yes, you're correct.